source/agora/common/BlockStorage.d

```diff
@@ -159,35 +298,346 @@ public class BlockStorage
+
+    ***************************************************************************/
+
+    private ubyte[] read (size_t from, size_t to) @trusted
+    {
+        assert(this.file !is null);
+        assert(this.getFileLength >= to);
+
+        return cast(ubyte[])this.file[from .. to];
+    }
+
+    /***************************************************************************
+
+        Write data to the file.
+
+        Params:
+            pos  = Start position of range to write
+            data = Array of unsigned bytes to be written to file
+
+    ***************************************************************************/
+
+    private void write (size_t pos, const ubyte[] data) @trusted
+    {
+        assert(this.file !is null);
+        assert(this.getFileLength >= pos+data.length);
+
+        foreach (idx, ref e; data)
+            this.file[pos + idx] = e;
+    }
+
+    /***************************************************************************
+
+        Get memory mapped file size
+
+        Returns:
+            Returns size of file if is mapped, or 0 if is not mapped
+
+    ***************************************************************************/
+
+    private size_t getFileLength () @trusted
+    {
+        if (this.file is null)
+            return 0;
+
+        return this.file.length;
+    }
+
+    /***************************************************************************
+
+        Store index data for one block in the file.
+
+        Params:
+            height = height of `Block`
+            hash = hash of `Block`
+            pos = position of memory mapped file
+
+        Returns:
+            Returns size of file if is mapped, or 0 if is not mapped
+
+    ***************************************************************************/
+
+    private void saveIndex (
+        size_t height,
+        ubyte[HASH_SIZE] hash,
+        size_t pos) @safe
+    {
+        File idx_file;
+        string file_name = buildPath(this.path, "index.dat");
+
+        try
+        {
+            idx_file = File(file_name, "a+b");
+            idx_file.seek(0, SEEK_END);
+
+            idx_file.writeSizeType(height);
+            idx_file.rawWrite(hash);
+            idx_file.writeSizeType(pos);
+
+            idx_file.close();
+        }
+        catch (Exception ex)
+        {
+            logError("saveIndex: %s",
+                () @trusted { return ex.message; }());
+        }
+    }
+
+    /***************************************************************************
+
+        Read the index data stored in the index file.
+
+    ***************************************************************************/
+
+    private void loadAllIndexes () @safe
+    {
+        File idx_file;
+
+        size_t height;
+        ubyte[HASH_SIZE] hash;
+        size_t pos;
+        string file_name = buildPath(this.path, "index.dat");
+
+        this.height_idx.clear();
+        this.hash_idx.clear();
+
+        if (file_name.exists)
+        {
```

Comment:
Use early returns to reduce indentation.
In this case, instead of wrapping the rest of the function in a `if`, use:
```d
if (!file_name.exists())
    return;
```

This way the code is simpler to read.